File: EPI/8.8/swift/8.8.swift
public class Node {
    public var next: Node?

    public func removeLast(index: Int) -> Node? {
        var node = self
        for _ in 0..<index {
            node = node.next!
        }

        if node.next == nil {
            // The index-th last node is the first node.
            return self.next
        }

        node = node.next!
        var parentOfNodeToRemove = self

        while let next = node.next {
            node = next
            parentOfNodeToRemove = parentOfNodeToRemove.next!
        }

        parentOfNodeToRemove.next =  parentOfNodeToRemove.next!.next

        return self
    }
}

File: EPI/Linked Lists/Remove the k-th last element from a list/typescript/8.8.ts
export interface List {
    next?: List;
}

export function removeLast(list: List, k: number): List | undefined {
    let left: List | undefined;
    let right = list;

    for (let rightIndex = 0; right.next && rightIndex < k; rightIndex += 1) {
        right = right.next;
    }

    while (right.next) {
        left = left ? left.next : list;
        right = right.next;
    }

    if (left) {
        // The node to remove is not the head.
        left.next = left.next!.next;
        return list;
    }

    // Remove the head.
    return list.next;
}
